### PR TITLE
Ternary search code in c

### PR DESCRIPTION
**Pull request by salonishah01**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Ternary search code in c

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#1326

### CODE REVIEW

**Comment by salonishah01** [comment]
@atarax665 Please review this PR

**Review by atarax665 — CHANGES_REQUESTED** [review]
Add iterative approach too. And shift this file in ```searching``` folder and make change in README.md accordingly.

File: C/other/ternarySearch.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#include <stdio.h> 
+int ternarySearch(int l, int r, int key, int ar[]) 
+{ 
+    if (r >= l) {
```
Comment (Abhijit2505) [inline C/other/ternarySearch.c]: Please change the form of the variable name `r`,`l` etc to some good name, that will allow others to view a cleaner code.

File: C/other/ternarySearch.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#include <stdio.h>
```
Comment (Abhijit2505) [inline C/other/ternarySearch.c]: Better code has comments in it, as this method used recursion, try to include some comments as well, that will make it easier for a beginner to understand.

**Review by Abhijit2505 — CHANGES_REQUESTED** [review]
The code looks good to me, just a few necessary changes requested 😊

**Comment by salonishah01** [comment]
@atarax665 @Abhijit2505 Mentioned changes are done.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Format your code like this. Everything else looks perfect.

```
#include <stdio.h>

int ternarySearchIterative(int left, int right, int key, int array[])
{
	while (right >= left)
	{
		int mid1 = left + (right - left) / 3;
		int mid2 = right - (right - left) / 3;
		if (array[mid1] == key)
		{
			return mid1;
		}

		if (array[mid2] == key)
		{
			return mid2;
		}

		if (key < array[mid1])
		{
			right = mid1 - 1;
		}
		else if (key > array[mid2])
		{
			left = mid2 + 1;
		}
		else
		{
			left = mid1 + 1;
			right = mid2 - 1;
		}
	}

	return -1;
}

int ternarySearchRecursive(int left, int right, int key, int array[])
{
	// Find the mid1 and mid2 
	if (right >= left)
	{
		int mid1 = left + (right - left) / 3;
		int mid2 = right - (right - left) / 3;
		// Check if key is present at any mid 
		if (array[mid1] == key)
		{
			return mid1;
		}

		if (array[mid2] == key)
		{
			return mid2;
		}

		// Since key is not present at mid, 
		// check in which region it is present 
		// then repeat the Search operation 
		// in that region 
		if (key < array[mid1])
		{
			// The key lies in between l and mid1 
			return ternarySearchRecursive(left, mid1 - 1, key, array);
		}
		else if (key > array[mid2])
		{
			// The key lies in between mid2 and r 
			return ternarySearchRecursive(mid2 + 1, right, key, array);
		}
		else
		{
			// The key lies in between mid1 and mid2 
			return ternarySearchRecursive(mid1 + 1, mid2 - 1, key, array);
		}
	}

	// Key not found 
	return -1;
}

int main()
{
	int num;
	printf("Enter number of elements: ");
	scanf("%d", &num);
	printf("\nEnter %d sorted elements: ", num);
	int array[num];
	for (int i = 0; i < num; i++)
	{
		scanf("%d", &array[i]);
	}

	int key, index;
	printf("\nEnter Key to be searched: ");
	scanf("%d", &key);
	printf("\nRecursive Search:");
	index = ternarySearchRecursive(0, num, key, array);
	printf("\nIndex of %d is %d.", key, index);
	printf("\nIterative Search:");
	index = ternarySearchIterative(0, num, key, array);
	printf("\nIndex of %d is %d.", key, index);
}

/*
Time Complexity: O(log3 n)
Space Complexity: O(1)

Input :
Enter number of elements: 8
Enter 8 elements: 12 25 48 52 67 78 89 91
Enter Key to be searched: 52

Output : 
Recursive Search: 
Index of 52 is 3.
Iterative Search:
Index of 52 is 3.
*/
```

**Comment by salonishah01** [comment]
@atarax665 Formatted the code accordingly.

**Review by atarax665 — APPROVED** [review]
LGTM :)

**Review by iamrajiv — APPROVED** [review]

**Pull request merged**